src/Mapper.php

```diff
@@ -1,11 +1,156 @@
+
+        if ($attributes = $instance->reflection->getAttributes(CaseConversionInterface::class, ReflectionAttribute::IS_INSTANCEOF)) {
+            if (count($attributes) > 1) {
+                throw new MapperException("{$class} has more than one case conversion attribute");
+            }
+
+            $instance->caseConversion = $attributes[0]->newInstance();
+        }
+
+        $instance->instance = new $class();
+
+        return $instance->doMapping();
+    }
+
+    protected function convertNameToCase(string $name): string
+    {
+        return $this?->caseConversion?->convertToCase($name) ?? $name;
+    }
+
+    protected function convertNameFromCase(string $name): string
+    {
+        return $this?->caseConversion?->convertFromCase($name) ?? $name;
+    }
+
+    protected function doMapping(): object
+    {
+        $this->fillPropertyAttributes();
+
+        foreach ($this->attributes as $types) {
+            $attribute = $types[0];
+            $data = $this->data[$this->convertNameFromCase($attribute->name)] ?? null;
+
+            if ($data === null) {
+                if ($attribute->nullable) {
+                    continue;
+                }
+
+                throw new MapperException("Property {$attribute->name} is not nullable");
+            }
+
+            $validType = false;
+
+            foreach ($types as $attribute) {
+                if ($attribute instanceof ObjectType) {
+                    $data = $this->map($attribute->class, $data);
+                }
+
+                if ($attribute instanceof ObjectArrayType) {
+                    $data = array_map(
+                        fn ($item) => $this->map($attribute->class, $item),
+                        $data
+                    );
+                }
+
+                if ($attribute->isType($data)) {
+                    $validType = true;
+                    break;
+                }
+            }
+
+            if (!$validType) {
+                throw new MapperException(
+                    "Property {$attribute->name} is not of type ".
+                    implode(
+                        ', ',
+                        array_map(static fn ($type) => $type::class, $types)
+                    )
+                );
+            }
+
+            $this->instance->{$this->convertNameToCase($attribute->name)} = $data;
+        }
+
+        return $this->instance;
+    }
+
+    protected function fillPropertyAttributes(): void
+    {
+        $properties = $this->reflection->getProperties(ReflectionProperty::IS_PUBLIC);
+
+        foreach ($properties as $property) {
+            $attributes = array_map(
+                static fn (ReflectionAttribute $attribute) => $attribute->newInstance(),
+                $property->getAttributes(BaseType::class, ReflectionAttribute::IS_INSTANCEOF)
+            );
+
+            $name = $property->getName();
+            $this->attributes[$name] = [];
+
+            if ($attributes !== []) {
```

Comment:
`if (!empty($attributes))`

Reply:
Comparing to empty array for speed

Reply:
That makes no sense, `!empty()` is faster. See: https://onlinephp.io/c/37ad5